WIP [11.0][ADD][sale_order_project] Initial migration attempt

SalahAdDin opened the pull request:
I'm doing this from 10.0 module's version, i added Turkish translation and also i'm updating some things bassed in `crm_project` module.
If you rebase this branch from the main repo all difference related to the other modules will be eliminated. Thank you very much.

SalahAdDin committed:
Merge pull request #1 from OCA/11.0

Update from original

SalahAdDin committed:
WIP [11.0][ADD][sale_order_project] Initial migration attempt

SalahAdDin committed:
:fire: applied 2to3 command.

SalahAdDin committed:
:sunny: improved spanish translation and added turkish translation.

SalahAdDin committed:
:fire: remove blank lines

aitorbouzas commented:
Just updated the forked branch.

SalahAdDin commented:
Ok, let me a minute i will continue working in this module.

SalahAdDin commented:
@aitorbouzas There is an importante change in `crm_project`, it does not use models, it use a view and wizard.

SalahAdDin committed:
Merge branch '11.0-mig-sale_order_project' into 11.0

SalahAdDin commented:
I have a problem with `related_project_id` in the view:
```bash
Error:
Odoo Server Error

Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/odoo/models.py", line 1036, in _validate_fields
    check(self)
  File "/usr/lib/python3/dist-packages/odoo/addons/base/ir/ir_ui_view.py", line 330, in _check_xml
    view_def = view.read_combined(['arch'])
  File "/usr/lib/python3/dist-packages/odoo/addons/base/ir/ir_ui_view.py", line 720, in read_combined
    arch = self.apply_view_inheritance(arch_tree, root.id, self.model)
  File "/usr/lib/python3/dist-packages/odoo/addons/base/ir/ir_ui_view.py", line 669, in apply_view_inheritance
    source = self.apply_inheritance_specs(source, specs_tree, view_id)
  File "/usr/lib/python3/dist-packages/odoo/addons/base/ir/ir_ui_view.py", line 647, in apply_inheritance_specs
    self.raise_view_error(_("Element '%s' cannot be located in parent view") % tag, inherit_id)
  File "/usr/lib/python3/dist-packages/odoo/addons/base/ir/ir_ui_view.py", line 509, in raise_view_error
    raise ValueError(message)
ValueError: Element '<field name="related_project_id">' cannot be located in parent view

Error context:
View `view_project_quotation_order_form`
[view_id: 1714, xml_id: n/a, model: sale.order, parent_id: 1213]

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/odoo/tools/convert.py", line 744, in parse
    self._tags[rec.tag](rec, de, mode=mode)
  File "/usr/lib/python3/dist-packages/odoo/tools/convert.py", line 654, in _tag_record
    id = self.env(context=rec_context)['ir.model.data']._update(rec_model, self.module, res, rec_id or False, not self.isnoupdate(data_node), noupdate=self.isnoupdate(data_node), mode=self.mode)
  File "/usr/lib/python3/dist-packages/odoo/addons/base/ir/ir_model.py", line 1482, in _update
    record = record.create(values)
  File "/usr/lib/python3/dist-packages/odoo/addons/base/ir/ir_ui_view.py", line 398, in create
    return super(View, self).create(self._compute_defaults(values))
  File "/usr/lib/python3/dist-packages/odoo/models.py", line 3381, in create
    self._fields[key].determine_inverse(record)
  File "/usr/lib/python3/dist-packages/odoo/fields.py", line 1071, in determine_inverse
    getattr(records, self.inverse)()
  File "/usr/lib/python3/dist-packages/odoo/addons/base/ir/ir_ui_view.py", line 266, in _inverse_arch
    view.write(data)
  File "/usr/lib/python3/dist-packages/odoo/addons/base/ir/ir_ui_view.py", line 414, in write
    return super(View, self).write(self._compute_defaults(vals))
  File "/usr/lib/python3/dist-packages/odoo/models.py", line 3090, in write
    self._write(old_vals)
  File "/usr/lib/python3/dist-packages/odoo/models.py", line 3256, in _write
    self._validate_fields(vals)
  File "/usr/lib/python3/dist-packages/odoo/models.py", line 1040, in _validate_fields
    raise ValidationError("%s\n\n%s" % (_("Error while validating constraint"), tools.ustr(e)))
odoo.exceptions.ValidationError: ('Error while validating constraint\n\nElement \'<field name="related_project_id">\' cannot be located in parent view\n\nError context:\nView `view_project_quotation_order_form`\n[view_id: 1714, xml_id: n/a, model: sale.order, parent_id: 1213]', None)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/odoo/http.py", line 650, in _handle_exception
    return super(JsonRequest, self)._handle_exception(exception)
  File "/usr/lib/python3/dist-packages/odoo/http.py", line 310, in _handle_exception
    raise pycompat.reraise(type(exception), exception, sys.exc_info()[2])
  File "/usr/lib/python3/dist-packages/odoo/tools/pycompat.py", line 87, in reraise
    raise value
  File "/usr/lib/python3/dist-packages/odoo/http.py", line 692, in dispatch
    result = self._call_function(**self.params)
  File "/usr/lib/python3/dist-packages/odoo/http.py", line 342, in _call_function
    return checked_call(self.db, *args, **kwargs)
  File "/usr/lib/python3/dist-packages/odoo/service/model.py", line 97, in wrapper
    return f(dbname, *args, **kwargs)
  File "/usr/lib/python3/dist-packages/odoo/http.py", line 335, in checked_call
    result = self.endpoint(*a, **kw)
  File "/usr/lib/python3/dist-packages/odoo/http.py", line 936, in __call__
    return self.method(*args, **kw)
  File "/usr/lib/python3/dist-packages/odoo/http.py", line 515, in response_wrap
    response = f(*args, **kw)
  File "/usr/lib/python3/dist-packages/odoo/addons/web/controllers/main.py", line 935, in call_button
    action = self._call_kw(model, method, args, {})
  File "/usr/lib/python3/dist-packages/odoo/addons/web/controllers/main.py", line 923, in _call_kw
    return call_kw(request.env[model], method, args, kwargs)
  File "/usr/lib/python3/dist-packages/odoo/api.py", line 689, in call_kw
    return call_kw_multi(method, model, args, kwargs)
  File "/usr/lib/python3/dist-packages/odoo/api.py", line 680, in call_kw_multi
    result = method(recs, *args, **kwargs)
  File "<decorator-gen-65>", line 2, in button_immediate_install
  File "/usr/lib/python3/dist-packages/odoo/addons/base/module/module.py", line 71, in check_and_log
    return method(self, *args, **kwargs)
  File "/usr/lib/python3/dist-packages/odoo/addons/base/module/module.py", line 448, in button_immediate_install
    return self._button_immediate_function(type(self).button_install)
  File "/usr/lib/python3/dist-packages/odoo/addons/base/module/module.py", line 541, in _button_immediate_function
    modules.registry.Registry.new(self._cr.dbname, update_module=True)
  File "/usr/lib/python3/dist-packages/odoo/modules/registry.py", line 85, in new
    odoo.modules.load_modules(registry._db, force_demo, status, update_module)
  File "/usr/lib/python3/dist-packages/odoo/modules/loading.py", line 380, in load_modules
    loaded_modules, update_module, models_to_check)
  File "/usr/lib/python3/dist-packages/odoo/modules/loading.py", line 274, in load_marked_modules
    perform_checks=perform_checks, models_to_check=models_to_check
  File "/usr/lib/python3/dist-packages/odoo/modules/loading.py", line 181, in load_module_graph
    _load_data(cr, module_name, idref, mode, kind='data')
  File "/usr/lib/python3/dist-packages/odoo/modules/loading.py", line 95, in _load_data
    tools.convert_file(cr, module_name, filename, idref, mode, noupdate, kind, report)
  File "/usr/lib/python3/dist-packages/odoo/tools/convert.py", line 791, in convert_file
    convert_xml_import(cr, module, fp, idref, mode, noupdate, report)
  File "/usr/lib/python3/dist-packages/odoo/tools/convert.py", line 852, in convert_xml_import
    obj.parse(doc.getroot(), mode=mode)
  File "/usr/lib/python3/dist-packages/odoo/tools/convert.py", line 751, in parse
    exc_info[2]
  File "/usr/lib/python3/dist-packages/odoo/tools/pycompat.py", line 86, in reraise
    raise value.with_traceback(tb)
  File "/usr/lib/python3/dist-packages/odoo/tools/convert.py", line 744, in parse
    self._tags[rec.tag](rec, de, mode=mode)
  File "/usr/lib/python3/dist-packages/odoo/tools/convert.py", line 654, in _tag_record
    id = self.env(context=rec_context)['ir.model.data']._update(rec_model, self.module, res, rec_id or False, not self.isnoupdate(data_node), noupdate=self.isnoupdate(data_node), mode=self.mode)
  File "/usr/lib/python3/dist-packages/odoo/addons/base/ir/ir_model.py", line 1482, in _update
    record = record.create(values)
  File "/usr/lib/python3/dist-packages/odoo/addons/base/ir/ir_ui_view.py", line 398, in create
    return super(View, self).create(self._compute_defaults(values))
  File "/usr/lib/python3/dist-packages/odoo/models.py", line 3381, in create
    self._fields[key].determine_inverse(record)
  File "/usr/lib/python3/dist-packages/odoo/fields.py", line 1071, in determine_inverse
    getattr(records, self.inverse)()
  File "/usr/lib/python3/dist-packages/odoo/addons/base/ir/ir_ui_view.py", line 266, in _inverse_arch
    view.write(data)
  File "/usr/lib/python3/dist-packages/odoo/addons/base/ir/ir_ui_view.py", line 414, in write
    return super(View, self).write(self._compute_defaults(vals))
  File "/usr/lib/python3/dist-packages/odoo/models.py", line 3090, in write
    self._write(old_vals)
  File "/usr/lib/python3/dist-packages/odoo/models.py", line 3256, in _write
    self._validate_fields(vals)
  File "/usr/lib/python3/dist-packages/odoo/models.py", line 1040, in _validate_fields
    raise ValidationError("%s\n\n%s" % (_("Error while validating constraint"), tools.ustr(e)))
odoo.tools.convert.ParseError: "Error while validating constraint

Element '<field name="related_project_id">' cannot be located in parent view

Error context:
View `view_project_quotation_order_form`
[view_id: 1714, xml_id: n/a, model: sale.order, parent_id: 1213]
None" while parsing /var/lib/odoo/addons/11.0/project/sale_order_project/views/sale_view.xml:3, near
<record id="view_order_form" model="ir.ui.view">
        <field name="name">view_project_quotation_order_form</field>
        <field name="model">sale.order</field>
        <field name="inherit_id" ref="sale.view_order_form"/>
        <field name="arch" type="xml">
            <field name="related_project_id" position="after">
                <button name="action_create_project" type="object" string="Create Project" class="oe_inline oe_edit_only" groups="sale.group_analytic_accounting,project.group_project_user" attrs="{'invisible':['|',('state','!=','draft'),('related_project_id','!=',False)]}"/>
            </field>
        </field>
    </record>
```

aitorbouzas commented:
Well, related_project_id is a field I created in that same module (sale_order_project), try not to use it to position your button, use another field of view_order_form.

SalahAdDin committed:
:fire: improve manifest file.

SalahAdDin commented:
@aitorbouzas what do you think about using the `crm_project` way? it does not use models, just wizards.
Ok, it is not the same thing, that wizard converts a lead in a project's task, we want to create or choose a project and linking both.

SalahAdDin committed:
:fire: Refactored `Create Project` button.

SalahAdDin committed:
:fire: Removed deprecated license lines.

SalahAdDin committed:
:fire: Removed deprecated license lines.

SalahAdDin committed:
:sunny: Fix button ihneritance.

SalahAdDin committed:
:fire: Removed deprecated license lines.

SalahAdDin committed:
:fire: Reverted to the local approach.

SalahAdDin commented:
Ok, i could install it finally, i'm using demo data for testing purposes, but I still have some problems: 
First one, the button does not appear in all existing quotations, it appears at least in two quotations:
![aquiotravez](https://user-images.githubusercontent.com/5034215/43675061-843f7d10-97e6-11e8-99cd-0aa69036579e.png)
![auisiaparece](https://user-images.githubusercontent.com/5034215/43675062-848a1ece-97e6-11e8-918f-2871004bb11e.png)
In the correct place but with wrong icon, i will fix it right now; but it does not appears in other:
![aquino](https://user-images.githubusercontent.com/5034215/43675068-a5a6710c-97e6-11e8-9615-1aeda0f46865.png)
It appears in new quotations and you can click it after fulfill the other fields:
![soloapareceennuevasordenes](https://user-images.githubusercontent.com/5034215/43675081-c01a7ef2-97e6-11e8-82a0-98ec2176f794.png)

The second one, if we do click in the button, it does not show the pop up dialog, I'm getting this bug in the browser's console:
```
No type for action Object
do_action @ action_manager.js:681
```

And, of course, the related project field  does not appears in the field's group.

SalahAdDin committed:
:fire: Fixed button icon.

SalahAdDin committed:
:fire: Reverted button type to  Object.

I'm sorry, it was a noob mistake.

SalahAdDin committed:
:fixed: sales.order has not project_id attribute.

SalahAdDin commented:
@aitorbouzas 
Ok, for now i think this is enough:
![deepin-screen-recorder_select area_20180806090945](https://user-images.githubusercontent.com/5034215/43699786-d88706a4-9958-11e8-8c3f-cc09bc8ef825.gif)
Unfortunately, it does not show the create project pop-up dialog, it creates a new project automatically. It is enough for my client's requirements, but i think with a pop-up dialog could be great.

can you review this pull request, please?

Thank you.

aitorbouzas commented:
What kind of popup do you want to show? A confirmation button?
Anyway, I think the "Create project" button would be better if it was located in the header of the button instead of the button box.

aitorbouzas commented on sale_order_project/views/sale_view.xml:
```diff
@@ -1,24 +1,33 @@
+                        icon="fa-tasks"
+                        help="Create Project"
+                        groups="sale.group_analytic_accounting,project.group_project_user"
+                        attrs="{'invisible':['|',('state','!=','draft'),('related_project_id','!=',False)]}">
```
If you want a confirmation add a new attribute: ``` confirm="Text to show on the confirm dialog". ```

SalahAdDin replied:
![deepin-screen-recorder_select area_20180806100950](https://user-images.githubusercontent.com/5034215/43702026-1529ab7c-9961-11e8-9c26-9fd9ea60ae79.gif)
I'm thinking about something like this.

aitorbouzas replied:
Ok, you have to create a new wizard for that. The action of the button should return a act_window action with model of the wizard.

SalahAdDin replied:
In your opinion, which is the best option here: with a pop-up or with it is right now?

aitorbouzas replied:
If you don't want the user to input something (like the name of the project to create) I think that confirmation dialog is enough.

SalahAdDin replied:
I prefer this approach:
> In a manufacturing order (MO), you can select a project to be attached to it. If none is selected, a project is automatically created when the MO is confirmed.

It is the [`mrp_project`](https://github.com/OCA/manufacture/tree/8.0/mrp_project) module.

I mean, if the user wants to attach an existing project user can choose, but if the user does not choose anything, the current behavior is the ideal.

aitorbouzas replied:
Ok, do you know how to make this?

SalahAdDin replied:
I will check it.

SalahAdDin replied:
@aitorbouzas I'm not sure about how can i create a new wizard, i made the form, but i don't know how to do the other things.

Anyway, i'm searching about that.

aitorbouzas replied:
The button should return an action or call a method that returns an action. This is an example of a button that returns an action:
```
<button class="oe_highlight" attrs="{'invisible': [('sale_number','=', 0)]}" type="action"
                        string="Make certification(s)"
                        name="%(crm_lead_make_invoice_action)d" />
```

And this is the action that shows the popup form:

```
<act_window name="Make invoice from lead" id="crm_lead_make_invoice_action"
                res_model="crm.lead.make_invoice"
                view_mode="form" target="new"/>
```

You also have to have a form view for the wizard model ```crm.lead.make_invoice``` that will define the wizard view.

SalahAdDin replied:
Yes, i did all of it, but the `python` code is not so easy, i think. I will up a code but i still does not work.

SalahAdDin replied:
@aitorbouzas Can you review my recent commit? please.

aitorbouzas commented on sale_order_project/views/sale_view.xml:
```diff
@@ -1,24 +1,33 @@
+                    This must to be the proper icon, but, until Odoos 12 version the current Font-Awesome version is 4.7, such has not this icon; instead we will use the current icon.
+                    https://github.com/odoo/odoo/issues/21648#issuecomment-353099417
+                -->
+                <button class="oe_stat_button"
```
Add a new position for this button. I think it is better to have buttons that creates registers and make changes in the active id on the header. The button box is a nice way to have only buttons to check related models.

SalahAdDin replied:
Well, as you can see `create` button becomes `Project(s) Overview` after to create the new project.

Where must i to put this button?

aitorbouzas replied:
After "action_cancel" button or so. Just make sure you make it invisible if the state is not confirmed.

SalahAdDin replied:
Mmmm, but `crm_project` module use this button in the button box.

aitorbouzas replied:
Can't find crm_project module, is it an OCA one? Could you provide a link?

SalahAdDin replied:
It is in [Odoo](https://github.com/odoo/odoo/tree/11.0/addons/crm_project), and, you can see it in the above `gift`.

aitorbouzas replied:
Ok, I see. You can leave as it is right now if you want, but I still think that the buttons box should only be there for checking related models and activate/deactivate the active model.

SalahAdDin replied:
Ok, we can also get feedback from `OCA` users.

aitorbouzas replied:
Yup, as soon as we merge this my PR will be updated (https://github.com/OCA/project/pull/395) and we'll receive feedback (I guess).

SalahAdDin commented:
@aitorbouzas As far i read in the `Odoo 11 Development Essentials`, `api.one` is deprecated, and we must change it for `api.multi`:
```
# @api.one
    @api.multi
    @api.depends('analytic_account_id')
    def _compute_related_project_id(self):
        self.ensure_one()
        self.related_project_id = (
            self.env['project.project'].search(
                [('analytic_account_id', '=', self.analytic_account_id.id)],limit=1
            )[:1]
        )
```

This deprecation raises a warning in the shell that makes to the second's test fail.

aitorbouzas commented:
Yup, you are right on the api.one deprecation. Just use api.multi and self.ensure_one() or make a loop.

SalahAdDin committed:
:fire: Fix deprecation warning: api.one

SalahAdDin committed:
:tada: Initial wizard attempt.

aitorbouzas commented on sale_order_project/models/sale.py:
```diff
@@ -25,30 +23,3 @@ def _compute_related_project_id(self):
         string='Project',
         compute='_compute_related_project_id'
     )
-
```
You deleted all the methods from sale.order model. This is not a good practice since the wizard should use them.

SalahAdDin replied:
I put them in the `wizard`.

aitorbouzas commented on sale_order_project/wizard/sale_order_createproject.py:
```diff
@@ -0,0 +1,46 @@
+    related_project_id = fields.Many2one('project.project', string='Project')
+
+    @api.model
+    def _prepare_project_vals(self, order):
```
As I said in the other comment, is better if you leave this behaviour in sale.order so you don't have to worry about passing the order. The wizard should only call methods on sale.order or write over the account_analytic_id field to assign an already created project.

SalahAdDin replied:
So, what must i to put in the `wizard` file?

aitorbouzas replied:
The wizard python file must get the input of the user and then do one of two things:
1. Either create a new project calling the method of the sale order "action_create_project".
2. Or write the sale order and create the relation sale order - project.

SalahAdDin replied:
How can i know if the `project` field is empty at clicking in the `Accept` button?

SalahAdDin replied:
After the new commit, we get a beautiful pop-up:
![deepin-screen-recorder_select area_20180807102101](https://user-images.githubusercontent.com/5034215/43760742-b65a030a-9a2b-11e8-99c8-3e36ccff1aab.gif)

aitorbouzas replied:
Nice, have you managed how to know if a field is empty?
You just have to:
```if order.related_project_id:```

SalahAdDin replied:
I just was looking for but i cannot do it, one moment.
Are you sure about it? 
Please, see the two fields that i put in the `wizard` python file.

SalahAdDin replied:
```python
class SaleOrderCreateProject(models.TransientModel):
    """ wizard to create a Project from a Sale Order """
    _name = "sale.order.createproject"

    @api.model
    def default_get(self, fields):
        result = super(SaleOrderCreateProject, self).default_get(fields)
        sale_order_id = self.env.context.get('active_id')
        if sale_order_id:
            result['sale_order_id'] = sale_order_id
        return result

    sale_order_id = fields.Many2one('sale.order', string='Order', domain=[('type', '=', 'order')])
    related_project_id = fields.Many2one('project.project', string='Project')

    @api.multi
    def action_create_project_task(self):
        self.ensure_one()
        # get the lead to transform
        order = self.sale_order_id
        project = self.related_project_id

        if order.related_project_id:
        # if related_project_id is empty
            # create new project.project
            return order.action_create_project()
        else:
        # else
            # update sale.order.related_project_id with the selected project.project.id
            vals = {
                'related_project_id': self.related_project_id,
            }
            order.write(vals)
            return True
```

SalahAdDin replied:
It is not checking if the field is empty or not, i choose a project and it try to do `rder.action_create_project()`, so i get this error:
```bash
Error:
Odoo Server Error

Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/odoo/http.py", line 650, in _handle_exception
    return super(JsonRequest, self)._handle_exception(exception)
  File "/usr/lib/python3/dist-packages/odoo/http.py", line 310, in _handle_exception
    raise pycompat.reraise(type(exception), exception, sys.exc_info()[2])
  File "/usr/lib/python3/dist-packages/odoo/tools/pycompat.py", line 87, in reraise
    raise value
  File "/usr/lib/python3/dist-packages/odoo/http.py", line 692, in dispatch
    result = self._call_function(**self.params)
  File "/usr/lib/python3/dist-packages/odoo/http.py", line 342, in _call_function
    return checked_call(self.db, *args, **kwargs)
  File "/usr/lib/python3/dist-packages/odoo/service/model.py", line 97, in wrapper
    return f(dbname, *args, **kwargs)
  File "/usr/lib/python3/dist-packages/odoo/http.py", line 335, in checked_call
    result = self.endpoint(*a, **kw)
  File "/usr/lib/python3/dist-packages/odoo/http.py", line 936, in __call__
    return self.method(*args, **kw)
  File "/usr/lib/python3/dist-packages/odoo/http.py", line 515, in response_wrap
    response = f(*args, **kw)
  File "/usr/lib/python3/dist-packages/odoo/addons/web/controllers/main.py", line 935, in call_button
    action = self._call_kw(model, method, args, {})
  File "/usr/lib/python3/dist-packages/odoo/addons/web/controllers/main.py", line 923, in _call_kw
    return call_kw(request.env[model], method, args, kwargs)
  File "/usr/lib/python3/dist-packages/odoo/api.py", line 685, in call_kw
    method = getattr(type(model), name)
AttributeError: type object 'sale.order.createproject' has no attribute 'action_create_project'
```

aitorbouzas replied:
Because your ```order``` is an instance of sale.order.createproject instead of sale.order.

SalahAdDin replied:
It means i have to get a `sale.order` with the current `order.id` and then call the action, right?

SalahAdDin replied:
Done!

SalahAdDin committed:
:fire: reverted method to model file.

SalahAdDin committed:
:sunny: added wizard to init file.

SalahAdDin committed:
:fire: Commented inner code.

For now i don't know how can i handle it.

SalahAdDin committed:
:fire: Properly named modules.

SalahAdDin committed:
:fire: Properly named modules.

SalahAdDin committed:
:fire: correct action calling.

Until here all works fine, we can install it without problem.

We just need to handle the action.

aitorbouzas commented on sale_order_project/wizard/sale_order_createproject.py:
```diff
@@ -0,0 +1,36 @@
+        order = self.sale_order_id
+        project = self.related_project_id
+        """
+        # if related_project_id is empty
```
Here is where you can place:
```
if project and not order.related_project_id:
    # Within this if the user picked a project and the order does not have a related one
else:
    # Within this else are the other cases, here you can create it

SalahAdDin replied:
I did, please, review my other comments.

SalahAdDin replied:
With that conditional, it still tries to create a new project instead just update the order.

aitorbouzas replied:
To get sale_order_id correctly in default_get:
```
order = self.env['sale.order'].browse(self.env.context.get('active_id'))
if order:
    result.update({'sale_order_id': order.id})
```

Make sure that the sale_order_id from the wizard is a sale.order.

SalahAdDin replied:
```bash
Error:
Odoo Server Error

Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/odoo/http.py", line 650, in _handle_exception
    return super(JsonRequest, self)._handle_exception(exception)
  File "/usr/lib/python3/dist-packages/odoo/http.py", line 310, in _handle_exception
    raise pycompat.reraise(type(exception), exception, sys.exc_info()[2])
  File "/usr/lib/python3/dist-packages/odoo/tools/pycompat.py", line 87, in reraise
    raise value
  File "/usr/lib/python3/dist-packages/odoo/http.py", line 692, in dispatch
    result = self._call_function(**self.params)
  File "/usr/lib/python3/dist-packages/odoo/http.py", line 342, in _call_function
    return checked_call(self.db, *args, **kwargs)
  File "/usr/lib/python3/dist-packages/odoo/service/model.py", line 97, in wrapper
    return f(dbname, *args, **kwargs)
  File "/usr/lib/python3/dist-packages/odoo/http.py", line 335, in checked_call
    result = self.endpoint(*a, **kw)
  File "/usr/lib/python3/dist-packages/odoo/http.py", line 936, in __call__
    return self.method(*args, **kw)
  File "/usr/lib/python3/dist-packages/odoo/http.py", line 515, in response_wrap
    response = f(*args, **kw)
  File "/usr/lib/python3/dist-packages/odoo/addons/web/controllers/main.py", line 931, in call_kw
    return self._call_kw(model, method, args, kwargs)
  File "/usr/lib/python3/dist-packages/odoo/addons/web/controllers/main.py", line 923, in _call_kw
    return call_kw(request.env[model], method, args, kwargs)
  File "/usr/lib/python3/dist-packages/odoo/api.py", line 687, in call_kw
    return call_kw_model(method, model, args, kwargs)
  File "/usr/lib/python3/dist-packages/odoo/api.py", line 672, in call_kw_model
    result = method(recs, *args, **kwargs)
  File "/usr/lib/python3/dist-packages/odoo/models.py", line 3350, in create
    vals = self._add_missing_default_values(vals)
  File "/usr/lib/python3/dist-packages/odoo/models.py", line 1602, in _add_missing_default_values
    for name, value in defaults.items():
AttributeError: 'NoneType' object has no attribute 'items'
```

SalahAdDin replied:
One, moment, i made a mistake in the `wizard` button. 
But, i have the same problem.

SalahAdDin replied:
No information in web, i cannot understand it.

aitorbouzas replied:
Check `sale` module wizards. You have one called `sale.make.invoice.advance`. Try to comprehend it completely and then compare it to yours. I'm not able to clone your fork and test it right now, sorry.

SalahAdDin committed:
:fire: fixed button action.

SalahAdDin committed:
:fire: refoactored default get method.

SalahAdDin committed:
:sunny: added a success message and trying to handle the action.

SalahAdDin committed:
:fire: tweaks in wizard.

-  Reverted default get change.
- Added help text to the field.
- Added conditionals.

_Note:_ Right know only the first condition works: it does not matter if you choose a project, it ever calls the create project method.

SalahAdDin committed:
:fire: Refactored field format in python 3 way.

SalahAdDin committed:
:fire: Removed useless pop up.

SalahAdDin committed:
:fire: Format tweaks.

SalahAdDin committed:
:fire: refactored conditionals, 

Right now the method can link sale order and project succesfully.
But for empty field it ever takes the `else` stament. Why?

SalahAdDin commented:
At this point, It just needs to filter correctly with a blank field, but i'm getting this from the field: 
`Warning: This sale order already has a related project. Project: project.project(), Order: sale.order(25,)`.

SalahAdDin commented:
I made this change:
```
    def action_create_project_task(self):
        self.ensure_one()
        # get the order to update
        order = self.sale_order_id
        project = self.related_project_id

        # if related_project_id is empty
        if not project.id and not order.related_project_id:
            # create new project.project
            order.action_create_project()
        # if project but order is not related to that project
        if project.id and not order.related_project_id:
            # update sale.order.related_project_id with the selected project.project.id
            order.write({
                'analytic_account_id': project.analytic_account_id.id
            })
        # else
        else:
            raise Warning(_(
                'This sale order already has a related project. Project: {0}, Order: {1}, Order related field: {2}'.format(project.id,order, order.related_project_id)
            ))
```

But it says that ok, we have not a `project.id` but we have a `order.related_project_id`, why?

SalahAdDin committed:
:fire: fixed conditionals

Four hours in a simple `elif`.

SalahAdDin commented:
@aitorbouzas Two things leave yet:
- [ ] Don't show the button when a Quotation is being created.
- [ ] Show quotation(s) in the Project description.

SalahAdDin committed:
:fix: Refactored warnings' format.

SalahAdDin committed:
:fire: Update readme with usage section.

SalahAdDin commented:
@aitorbouzas Can you review this please?

aitorbouzas commented:
It's looking good. I'll try it as soon as possible and come back.

SalahAdDin committed:
:fire: Fixed company email.

SalahAdDin committed:
:tada: [WIP] Added button in Project to see all related Sale Orders

Create a new field for count.

SalahAdDin committed:
:sunny: Added button for project view.

SalahAdDin committed:
:sunny: Added button for project view. reference in manifest.

SalahAdDin committed:
:fire: beautify sale view template.

SalahAdDin commented:
I'm trying with show challenge, i have a button, i have templates, etc., but it shows me errors.

SalahAdDin commented:
I am using an approach similar to `Documents` in the `Project` view, you can see it in the code.

But, when i add the `sales_order_count` to the `Project`, all projects show me that they have `20` sale orders, it means, they are taking all sale orders as their, which is wrong; seeing deeply in the log i can see that `sale.order.related_project_id` is not a stored field and that's why `osv` cannot search with it.

`odoo.osv.expression: Non-stored field sale.order.related_project_id cannot be searched.`

I cannot understand why.

SalahAdDin committed:
:sunny: Added Sales Management as dependencie.

It has no sense to use buttons and others without the proper view.

SalahAdDin committed:
:sunny: Added properly action to the button.

SalahAdDin committed:
:sunny: [WIP] Creating the action for button.

SalahAdDin commented:
One of this solutions makes that computer file stored, but, I think it could be not a proper way.

SalahAdDin commented:
In addition, i'm getting this strange behavior:
![deepin-screen-recorder_select area_20180810173808](https://user-images.githubusercontent.com/5034215/43963944-6fbea54a-9cc4-11e8-95a5-63468743dcba.gif)

SalahAdDin commented:
I'm investigating that. I tried with the first option: putting `store=True` in the field:
```
related_project_id = fields.Many2one(
        comodel_name='project.project',
        string='Project',
        compute='_compute_related_project_id',
        store=True
    )
```
But, i'm getting this bug:
```
Error:
Odoo Server Error

Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/odoo/fields.py", line 937, in __get__
    value = record.env.cache.get(record, self)
  File "/usr/lib/python3/dist-packages/odoo/api.py", line 960, in get
    value = self._data[field][record.id][key]
KeyError: <odoo.api.Environment object at 0x7f2cfae8fba8>

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/odoo/http.py", line 650, in _handle_exception
    return super(JsonRequest, self)._handle_exception(exception)
  File "/usr/lib/python3/dist-packages/odoo/http.py", line 310, in _handle_exception
    raise pycompat.reraise(type(exception), exception, sys.exc_info()[2])
  File "/usr/lib/python3/dist-packages/odoo/tools/pycompat.py", line 87, in reraise
    raise value
  File "/usr/lib/python3/dist-packages/odoo/http.py", line 692, in dispatch
    result = self._call_function(**self.params)
  File "/usr/lib/python3/dist-packages/odoo/http.py", line 342, in _call_function
    return checked_call(self.db, *args, **kwargs)
  File "/usr/lib/python3/dist-packages/odoo/service/model.py", line 97, in wrapper
    return f(dbname, *args, **kwargs)
  File "/usr/lib/python3/dist-packages/odoo/http.py", line 335, in checked_call
    result = self.endpoint(*a, **kw)
  File "/usr/lib/python3/dist-packages/odoo/http.py", line 936, in __call__
    return self.method(*args, **kw)
  File "/usr/lib/python3/dist-packages/odoo/http.py", line 515, in response_wrap
    response = f(*args, **kw)
  File "/usr/lib/python3/dist-packages/odoo/addons/web/controllers/main.py", line 938, in call_button
    action = self._call_kw(model, method, args, {})
  File "/usr/lib/python3/dist-packages/odoo/addons/web/controllers/main.py", line 926, in _call_kw
    return call_kw(request.env[model], method, args, kwargs)
  File "/usr/lib/python3/dist-packages/odoo/api.py", line 689, in call_kw
    return call_kw_multi(method, model, args, kwargs)
  File "/usr/lib/python3/dist-packages/odoo/api.py", line 680, in call_kw_multi
    result = method(recs, *args, **kwargs)
  File "<decorator-gen-58>", line 2, in button_immediate_upgrade
  File "/usr/lib/python3/dist-packages/odoo/addons/base/module/module.py", line 71, in check_and_log
    return method(self, *args, **kwargs)
  File "/usr/lib/python3/dist-packages/odoo/addons/base/module/module.py", line 603, in button_immediate_upgrade
    return self._button_immediate_function(type(self).button_upgrade)
  File "/usr/lib/python3/dist-packages/odoo/addons/base/module/module.py", line 542, in _button_immediate_function
    modules.registry.Registry.new(self._cr.dbname, update_module=True)
  File "/usr/lib/python3/dist-packages/odoo/modules/registry.py", line 85, in new
    odoo.modules.load_modules(registry._db, force_demo, status, update_module)
  File "/usr/lib/python3/dist-packages/odoo/modules/loading.py", line 376, in load_modules
    force, status, report, loaded_modules, update_module, models_to_check)
  File "/usr/lib/python3/dist-packages/odoo/modules/loading.py", line 274, in load_marked_modules
    perform_checks=perform_checks, models_to_check=models_to_check
  File "/usr/lib/python3/dist-packages/odoo/modules/loading.py", line 153, in load_module_graph
    registry.init_models(cr, model_names, {'module': package.name})
  File "/usr/lib/python3/dist-packages/odoo/modules/registry.py", line 314, in init_models
    models[0].recompute()
  File "/usr/lib/python3/dist-packages/odoo/models.py", line 4905, in recompute
    vals = {n: rec[n] for n in ns}
  File "/usr/lib/python3/dist-packages/odoo/models.py", line 4905, in <dictcomp>
    vals = {n: rec[n] for n in ns}
  File "/usr/lib/python3/dist-packages/odoo/models.py", line 4759, in __getitem__
    return self._fields[key].__get__(self, type(self))
  File "/usr/lib/python3/dist-packages/odoo/fields.py", line 941, in __get__
    self.determine_value(record)
  File "/usr/lib/python3/dist-packages/odoo/fields.py", line 1028, in determine_value
    self.compute_value(recs)
  File "/usr/lib/python3/dist-packages/odoo/fields.py", line 1008, in compute_value
    self._compute_value(records)
  File "/usr/lib/python3/dist-packages/odoo/fields.py", line 999, in _compute_value
    getattr(records, self.compute)()
  File "/var/lib/odoo/addons/11.0/project/sale_order_project/models/sale.py", line 16, in _compute_related_project_id
    self.ensure_one()
  File "/usr/lib/python3/dist-packages/odoo/models.py", line 4367, in ensure_one
    raise ValueError("Expected singleton: %s" % self)
ValueError: Expected singleton: sale.order(21, 8, 7, 6, 4, 3, 14, 9, 15, 10, 16, 11, 17, 12, 20, 19, 5, 2, 1, 18, 13)
```

aitorbouzas commented:
That's because of your `_compute_related_project_id` method. Instad of using `self.ensure_one()` you have to use a loop to run through the self recordset.
```
for record in self:
    # DO YOUR STUFF
```

SalahAdDin commented:
So, I'm trying to find another approach: using a new search function. 
So, I investigated the `search_count` function in the `Odoo` base module: https://github.com/odoo/odoo/blob/11.0/odoo/models.py#L1450
It consist of call the `search` function, which is a basic function and its core is the `_search` function.

So, I went to the `attachment` model and we can see that it use a custom `_search` function: https://github.com/odoo/odoo/blob/11.0/odoo/addons/base/ir/ir_attachment.py#L366

SalahAdDin commented:
@aitorbouzas In our logic, we just want to get the related project for that sale order, not for all.

aitorbouzas commented:
Sure, but the method is run for every recordset when you update the module in order to fill the value.

SalahAdDin commented:
What does it mean?

SalahAdDin commented:
I propose this:
```
project_obj = self.env['project.project']
        for record in self:
            project_domain = [('analytic_account_id', '=', record.analytic_account_id.id)]
            record.related_project_id = project_obj.search(project_domain, limit=1)[:1]
```

SalahAdDin commented:
Solved removing the line:
```python
@api.multi
    @api.depends('analytic_account_id')
    def _compute_related_project_id(self):
        # project_obj = self.env['project.project']
        for record in self:
            project_domain = [('analytic_account_id', '=', record.analytic_account_id.id)]
            record.related_project_id = (
                self.env['project.project'].search(project_domain, limit=1)[:1]
            )
```

SalahAdDin commented:
Ok, it does not work, count field does not take any record.

SalahAdDin committed:
:fire: Added function to sale order count field.

-> Related projec id is now stored in data base.
-> Button now show the related sale orders number.

SalahAdDin commented:
@aitorbouzas I solved that, the final problem is with the tree view, I don't understand that piece of code in project: https://github.com/odoo/odoo/blob/11.0/addons/project/models/project.py#L101

I think it needs to call a `sale.order` view, but i don't understand how.

SalahAdDin committed:
:fire: Fixed sale order tree view problem.

SalahAdDin commented:
Ok, Now we can see the quotations in the project view:
![deepin-screen-recorder_select area_20180815092953](https://user-images.githubusercontent.com/5034215/44135240-f4542e88-a070-11e8-9e91-65117c9e15c2.gif)
Unfortunately, we have no anymore the project overview button in the sale orders.
![nohayboton](https://user-images.githubusercontent.com/5034215/44135256-09545600-a071-11e8-87d3-91d9d957d0d7.png)

SalahAdDin commented:
@aitorbouzas It seems that test does not require refactoring, so, you can review it and merge.

And, i don't know how to update the localization files, so, if you update it i can update all translations. It seems the project review's button is working again.

SalahAdDin commented:
**Update:**
Fake button, it exists right now because of a conflict with another package that i'm testing: `mrp_project`.

TODO:

- [x] Show quotation(s) in the Project description.
- [ ] Refactor tests. 
- [ ] Review the Project Overview's button problem in Sale Order's view: it is missing.
- [x] Add a Sale Order's button in Account Analytic view (manufacture project module use it, it could be useful).

aitorbouzas commented:
Just invited you as a collaborator so you can push directly since I cannot review this lately. This way you can push and check your changes in the OCA PR. Let's hope someone can help you there.

aitorbouzas commented:
Before merging, please rebase (or squash) all of your commits.

SalahAdDin commented:
Ok, @aitorbouzas thank you. But I need someone that contribute with test and remake the language files.

aitorbouzas commented:
That can be done after you merge. Maybe someone can help you in the OCA PR.

SalahAdDin commented:
I will merge this because i have a problem with the `project_overview` button and i need help with it.
